### PR TITLE
Processes & Threads + Memory API + IPC + Services API

### PR DESCRIPTION
**Pull request by PixelyIon**
Reworks the API marginally and adds documentation to almost everything. In addition to these adds major features like working shared memory and private memory, also should support more than one thread now. Do not merge, Service Implementation is pending..

### CODE REVIEW

File: app/src/main/cpp/switch/kernel/service.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+#include "../nce.h"
+
+namespace lightSwitch::kernel::service {
+class Service {
+    Service();
+};
```
Comment (zephyren25) [inline app/src/main/cpp/switch/kernel/service.h]: You forgot the indentation here

File: app/src/main/cpp/switch/kernel/svc.h [hunk 2]
```diff
@@ -32,40 +28,99 @@ namespace lightSwitch::constant {
...
         // 5.0.0+
         constexpr uint8_t UserExceptionContextAddr = 0x14;
         // 6.0.0+
         constexpr uint8_t TotalMemoryAvailableWithoutMmHeap = 0x15;
         constexpr uint8_t TotalMemoryUsedWithoutMmHeap = 0x16;
     };
+    namespace status {
+        constexpr uint32_t success = 0x0; //!< "Success"
+        constexpr uint32_t unimpl = 0x177202; //!< "Unimplemented behaviour"
+    }
```
Comment (zephyren25) [inline app/src/main/cpp/switch/kernel/svc.h]: These error codes are also used in services, so you should probably consider moving them

File: app/src/main/cpp/switch/kernel/types/KObject.h [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+#include "../../common.h"
+
+namespace lightSwitch::kernel::type {
+    enum class KObjectType {
+        KThread, KProcess
+    };
+    class KObject {
+    public:
```
Comment (zephyren25) [inline app/src/main/cpp/switch/kernel/types/KObject.h]: Newline after enum declaration

File: app/src/main/cpp/switch/nce.cpp [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                ExecuteFunction(reinterpret_cast<void *>(UnmapSharedFunc), fregs, process.second->main_thread);
+                if (static_cast<int>(fregs.regs[0]) == -1)
+                    throw exception("An error occurred while unmapping shared region in child process");
+            }
+        }
+        int err = munmap(reinterpret_cast<void *>(address), size);
+        if (err == -1)
+            throw exception(fmt::format("An occurred while unmapping shared region: {}", strerror(errno)));
+        if (addr_memory_map.at(address).fd!=-1) {
+            if (close(addr_memory_map.at(address).fd)==-1)
```
Comment (zephyren25) [inline app/src/main/cpp/switch/nce.cpp]: Fix the spacing on the operators

File: app/src/main/cpp/switch/os.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    std::shared_ptr<type::KProcess> OS::CreateProcess(uint64_t address, size_t stack_size) {
+        auto *stack = static_cast<uint8_t *>(mmap(nullptr, stack_size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_NORESERVE | MAP_ANONYMOUS | MAP_STACK, -1, 0)); // NOLINT(hicpp-signed-bitwise)
+        if (stack == MAP_FAILED)
+            throw exception("Failed to allocate stack memory");
+        if (mprotect(stack, PAGE_SIZE, PROT_NONE)) {
+            munmap(stack, stack_size);
+            throw exception("Failed to create guard pages");
+        }
+        pid_t pid = clone(&ExecuteChild, stack + stack_size, CLONE_FS | SIGCHLD, nullptr); // NOLINT(hicpp-signed-bitwise)
+        if(pid==-1) throw exception(fmt::format("Call to clone() has failed: {}", strerror(errno)));
```
Comment (zephyren25) [inline app/src/main/cpp/switch/os.cpp]: Ditto

File: app/src/main/cpp/switch/os.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        std::shared_ptr<type::KProcess> process = std::make_shared<kernel::type::KProcess>(pid, address, reinterpret_cast<uint64_t>(stack), stack_size, state);
+        process_map[pid] = process;
+        state.logger->Write(Logger::DEBUG, "Successfully created process with PID: {}", pid);
+        return process;
+    }
+
+    void OS::KillThread(pid_t pid) {
+        auto process = process_map.at(pid);
+        if(process->main_thread==pid) {
```
Comment (zephyren25) [inline app/src/main/cpp/switch/os.cpp]: ditto

File: app/src/main/cpp/switch/common.h [hunk 7]
```diff
@@ -1,69 +1,284 @@
...
+
 
 namespace lightSwitch {
+    // Global typedefs
+    typedef std::runtime_error exception; //!< This is used as the default exception
+    typedef uint32_t handle_t; //!< The type of an handle
+
+    namespace constant {
+        // Memory
+        constexpr uint64_t base_addr = 0x8000000; //!< The address space base
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/common.h]: i'd say align your `=` signs throughout your code to make your code look nicer

File: app/src/main/cpp/switch/kernel/ipc.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
+#include <syslog.h>
+#include <cstdlib>
+#include "ipc.h"
+#include "types/KProcess.h"
+
+namespace lightSwitch::kernel::ipc {
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/kernel/ipc.cpp]: Change namespace

File: app/src/main/cpp/switch/common.h [hunk 9]
```diff
@@ -1,69 +1,284 @@
...
-#include "hw/cpu.h"
-#include "hw/memory.h"
+#include <sys/mman.h>
+#include <sys/ptrace.h>
+#include <cstdint>
+#include <stdexcept>
+#include <string>
+
 
 namespace lightSwitch {
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/common.h]: change namespace

File: app/src/main/cpp/switch/common.cpp [hunk 10]
```diff
@@ -3,13 +3,36 @@
 #include <syslog.h>
 
 namespace lightSwitch {
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/common.cpp]: change namespace

File: app/src/main/cpp/switch/kernel/ipc.h [hunk 11]
```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include <cstdint>
+#include <vector>
+#include "switch/common.h"
+
+namespace lightSwitch::kernel::ipc {
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/kernel/ipc.h]: change namespace

File: app/src/main/cpp/switch/kernel/ipc.h [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    struct CommandStruct {
+        // https://switchbrew.org/wiki/IPC_Marshalling#IPC_Command_Structure
+        uint16_t type : 16;
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/kernel/ipc.h]: i'd say align these `:` chunks throughout your code to make it look nicer

File: app/src/main/cpp/switch/kernel/service.cpp [hunk 13]
```diff
@@ -0,0 +1,7 @@
+#include "service.h"
+
+namespace lightSwitch::kernel::service {
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/kernel/service.cpp]: change namespace

File: app/src/main/cpp/switch/kernel/service.h [hunk 14]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+#include "../nce.h"
+
+namespace lightSwitch::kernel::service {
```
Comment (Xpl0itR) [inline app/src/main/cpp/switch/kernel/service.h]: i've said this a few times now so i wont repeat after this, change namespace here and in the rest of the files

**Comment by PixelyIon** [comment]
Milestone 2 introduces a new memory model that supports true shared memory with separate permissions for remote and local processes, implements svcQueryMemory and completes svcGetInfo further, adds IPC support with the IpcRequest and IpcResponse classes. This should be the last milestone before the full Services implementation :)

**Comment by zephyren25** [comment]
@PixelyIon remove the README from this PR if possible

File: app/src/main/cpp/switch/kernel/services/BaseService.h [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+
+      public:
+        BaseService(device_state &state) : state(state) {}
+
+        virtual const char *Name() = 0;
+
+        virtual ipc::IpcResponse HandleSyncRequest(ipc::IpcRequest &request) = 0;
+
+
+    };
```
Comment (zephyren25) [inline app/src/main/cpp/switch/kernel/services/BaseService.h]: Remove the extra newlines

File: app/src/main/cpp/switch/kernel/svc.cpp [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+        state.logger->Write(Logger::DEBUG, "Heap size was set to 0x{:X}", state.nce->GetRegister(wreg::w1));
+    }
+
+    void QueryMemory(device_state &state) {
+        Memory::MemoryInfo mem_inf;
+        u64 addr = state.nce->GetRegister(xreg::x2);
+        if (state.nce->memory_map.count(addr)) {
+            mem_inf = state.nce->memory_map.at(addr)->GetInfo(state.this_process->main_thread);
+        } else if (state.this_process->memory_map.count(addr)) {
+            mem_inf = state.this_process->memory_map.at(addr)->GetInfo();
```
Comment (zephyren25) [inline app/src/main/cpp/switch/kernel/svc.cpp]: Remove the extra brackets

File: app/src/main/cpp/switch/kernel/svc.cpp [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+        auto thread = state.this_process->CreateThread(state.nce->GetRegister(xreg::x1), state.nce->GetRegister(xreg::x2), state.nce->GetRegister(xreg::x3), static_cast<u8>(state.nce->GetRegister(wreg::w4)));
+        state.nce->SetRegister(wreg::w0, constant::status::success);
+        state.nce->SetRegister(wreg::w1, thread->handle);
+    }
+
+    void StartThread(device_state &state) {
+        auto &object = state.this_process->handle_table.at(static_cast<const unsigned int &>(state.nce->GetRegister(wreg::w0)));
+        if (object->type == type::KObjectType::KThread) {
+            std::static_pointer_cast<type::KThread>(object)->Start();
+        } else
```
Comment (zephyren25) [inline app/src/main/cpp/switch/kernel/svc.cpp]: Ditto

File: app/src/main/cpp/switch/kernel/svc.cpp [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+            state.nce->SetRegister(wreg::w1, std::static_pointer_cast<type::KThread>(object)->priority);
+        } else
+            throw exception("GetThreadPriority was called on a non-KThread object");
+    }
+
+    void SetThreadPriority(device_state &state) {
+        auto &object = state.this_process->handle_table.at(static_cast<const unsigned int &>(state.nce->GetRegister(wreg::w0)));
+        if (object->type == type::KObjectType::KThread) {
+            std::static_pointer_cast<type::KThread>(object)->Start();
+        } else
```
Comment (zephyren25) [inline app/src/main/cpp/switch/kernel/svc.cpp]: Ditto

File: app/src/main/cpp/switch/nce.cpp [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+        memory_map[item->address] = item;
+        memory_region_map[region] = item;
+        return item;
+    }
+
+    size_t NCE::GetSharedSize() {
+        size_t shared_size = 0;
+        for (auto &region : memory_map) {
+            shared_size += region.second->size;
+        }
```
Comment (zephyren25) [inline app/src/main/cpp/switch/nce.cpp]: Ditto

File: app/src/main/cpp/switch/os.cpp [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+    void OS::KillThread(pid_t pid) {
+        auto process = process_map.at(pid);
+        if (process->main_thread == pid) {
+            state.logger->Write(Logger::DEBUG, "Exiting process with PID: {}", pid);
+            // Erasing all shared_ptr instances to the process will call the destructor
+            // However, in the case these are not all instances of it we wouldn't want to call the destructor
+            for (auto&[key, value]: process->thread_map) {
+                process_map.erase(key);
+            };
```
Comment (zephyren25) [inline app/src/main/cpp/switch/os.cpp]: Ditto

File: app/src/main/cpp/switch/os.cpp [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+
+    void OS::SvcHandler(u16 svc, pid_t pid) {
+        this_process = process_map.at(pid);
+        this_thread = this_process->thread_map.at(pid);
+        if (svc::svcTable[svc]) {
+            state.logger->Write(Logger::DEBUG, "SVC called 0x{:X}", svc);
+            (*svc::svcTable[svc])(state);
+        } else {
+            throw exception(fmt::format("Unimplemented SVC 0x{:X}", svc));
+        }
```
Comment (zephyren25) [inline app/src/main/cpp/switch/os.cpp]: Ditto

**Comment by PixelyIon** [comment]
I locked the entire repository for a month, but I think it was worth it.

**Pull request merged**